File: Stack/Questions/921. Minimum Add to Make Parentheses Valid.java
class Solution {
    // Method to calculate the minimum number of parentheses to add to make the sequence valid
    public int minAddToMakeValid(String s) {
        int stack_size = 0;   // Variable to represent the size of the stack
        int miss_match = 0;   // Variable to count the mismatched parentheses

        for (char parentheses : s.toCharArray()) {
            if (parentheses == '(') {
                stack_size++;  // Increase the stack size when an opening parenthesis is encountered
            } else if (parentheses == ')' && stack_size > 0) {
                stack_size--;  // If a closing parenthesis is encountered and the stack is not empty, match with an opening parenthesis
            } else {
                miss_match++;  // If there is no matching opening parenthesis for a closing parenthesis, increase the mismatch count
            }
        }
        
        return stack_size + miss_match;  // Return the total number of parentheses to be added for a valid sequence
    }
}

File: Stack/Parentheses/921. Minimum Add to Make Parentheses Valid .kt
[Apporoach (Count Method)]
------------------------------------------------------------------------------------------
 
 fun minAddToMakeValid(s: String): Int {

        var countOpen:Int = 0;
        var countClose:Int = 0;

        s.forEach{parentheses ->
            if(parentheses == '('){
                    countOpen++;
            }
           else if(countOpen>0){
                    countOpen--;
            }
                else{               // for '))))' these case not '()' case
                    countClose++;
                }
            }

            return countOpen+countClose;

  }  

 
 
 
 [Apporoach (Stack Method)]
-----------------------------------------------------------------
 
 
 
  fun minAddToMakeValid(s: String): Int {
        val mainStack :Stack<Char> = Stack<Char>();

        s.forEach {parentheses ->
            if(!mainStack.empty()  && (mainStack.peek()=='(' && parentheses==')')){  // stack should not be empty
                mainStack.pop();

            }
            else {
                mainStack.push(parentheses);
            }
        }


        return mainStack.size;
        
    }
